### DISCUSSION TITLE
Wait for an event to accept responses?

### DISCUSSION THREAD

**Discussion by hjgodwin** [jspsych/jsPsych#2936] [General]
Greetings all, having great fun using jsPsych for making various experiments.

I'm working on a study at the moment that needs to wait for an event before participants can move to the next trial. Is there some way this is possible in jsPsych? To be clear, what I'm thinking of is this:

- Event happens, which then calls a function
- That function allows participants to move onto the the next trial. It probably will also change the display to say something like 'press space to move to the next trial' or similar.

I've had a look around using various options but can't seem to figure it out.

Thanks!

**Comment by ChristopheBossens** [comment]
Hi hjgodwin,

Could you go into a little bit more detail about what exactly it is you are trying to achieve? E.g., what is the participant seeing when you are waiting for an event, what type of event is it exactly, ... 

Since you mentioned events, JavaScript allows you to define custom events (similar to 'click', 'mousemove', ... events) that you can raise manually. In the code below I have created a custom event and the corresponding event handler simply logs some messages and finishes the current trial.

The keyboard response trial starts listening for this event. So you can either finish the trial by pressing a keyboard key. But, you can now also open the developer tools window in your browser, and in the console window raise the custom event using the following code:

`window.dispatchEvent(event)`

This will also finish the trial.

```
    const jsPsych = initJsPsych();

    // Create a custom test event and event handler function
    let event = new Event("custom_test");
    let eventHandler = function(event){
        console.log("The event has been raised");
        console.log(event);
        jsPsych.finishTrial();
    }
    

    const hello_trial = {
      type: jsPsychHtmlKeyboardResponse,
      stimulus: 'Hello world!',
      on_start: function(){
        // When the trial starts, start listening for the custom event
        console.log("Start listening for events");
        window.addEventListener("custom_test", eventHandler);
      },

      on_finish: function(data){
        // Executed by writing window.dispatchEvent(event) in the console.window
        console.log("Stop listening for events");
        console.log(data);
        window.removeEventListener("custom_test", eventHandler);
      }
    }

    jsPsych.run([hello_trial]); 
```

Is this something in the direction you are looking for?

**Reply by hjgodwin** [reply]
Thanks so much for this! It worked perfectly. I hadn't realised there was something so simple and useful as jsPsych.finishTrial() - what a fool I am!

Thanks again for all your help.
